# HG changeset patch
# Parent 2c5b47fc3ff0d81fcc3465fd4142f8da6baf4910
# User Tanvi Vyas <tvyas@mozilla.com>
Updated HUDService.jsm to alert about mixed content
* * *
Updated patch to account for refactor of webconsole.  Moved font color to the css files.  Alert not working yet because of issue with this.contentLocation.
* * *
Additional changes

diff --git a/browser/devtools/webconsole/HUDService.jsm b/browser/devtools/webconsole/HUDService.jsm
--- a/browser/devtools/webconsole/HUDService.jsm
+++ b/browser/devtools/webconsole/HUDService.jsm
@@ -2109,30 +2109,74 @@ HeadsUpDisplay.prototype = {
     let urlNode = this.chromeDocument.createElementNS(XUL_NS, "label");
     urlNode.setAttribute("crop", "center");
     urlNode.setAttribute("flex", "1");
     urlNode.setAttribute("title", request.url);
     urlNode.setAttribute("value", request.url);
     urlNode.classList.add("hud-clickable");
     urlNode.classList.add("webconsole-msg-body-piece");
     urlNode.classList.add("webconsole-msg-url");
+
+    let severity = SEVERITY_LOG;
+    let mixed = false;
+    let requestURIObject;
+    let contentWindowURI;
+
+    try{
+      let ioServ = Cc["@mozilla.org/network/io-service;1"].getService(Ci.nsIIOService);
+      requestURIObject = ioServ.newURI(request.url, null, null);
+      contentWindowURI = ioServ.newURI(this.contentLocation, null, null);
+      mixed = true;
+    } catch(e) {
+        /*
+         * TODO: catch NS_ERROR_MALFORMED_URI?
+         * if this error occurs, we skip the mixed content section and 
+         * the webconsole does not warn about mixed content
+         */
+    }
+
+
+    // if the operations to create nsIURI objects for the contentWindow and request succeeeded
+    if(mixed) {
+
+      //requestURIScheme is the scheme of the requested content - the entry in webconsole
+      //contentWindowURIScheme is the scheme of the current page. 
+      let requestURIScheme = requestURIObject.scheme;
+      let contentWindowURIScheme = contentWindowURI.scheme;
+
+      if(contentWindowURIScheme == "https" && requestURIScheme != "https") {
+          urlNode.classList.add("webconsole-mixed-content");
+
+          let mixedContentWarning = " ["+l10n.getStr("webConsoleMixedContentWarning")+"]";
+
+          /*TODO: take mixedContentWarning message and make it a link to a 
+             mdn page that explains mixed content.  How do I do this?
+            let mixedContentWarning = " [<a href='http://cisforcookie.org'>"+l10n.getStr
+                                        ("webConsoleMixedContentWarning")+"<a>]";
+          */
+          urlNode.setAttribute("value", request.url+mixedContentWarning);
+ 
+          severity = SEVERITY_WARNING;
+      }
+    }
+
     linkNode.appendChild(urlNode);
 
     let statusNode = this.chromeDocument.createElementNS(XUL_NS, "label");
     statusNode.setAttribute("value", "");
     statusNode.classList.add("hud-clickable");
     statusNode.classList.add("webconsole-msg-body-piece");
     statusNode.classList.add("webconsole-msg-status");
     linkNode.appendChild(statusNode);
 
     let clipboardText = request.method + " " + request.url;
 
     let messageNode = ConsoleUtils.createMessageNode(this.chromeDocument,
                                                      CATEGORY_NETWORK,
-                                                     SEVERITY_LOG,
+                                                     severity,
                                                      msgNode,
                                                      this.hudId,
                                                      null,
                                                      null,
                                                      clipboardText);
 
     messageNode.setAttribute("connectionId", entry.connection);
 
diff --git a/browser/locales/en-US/chrome/browser/devtools/webconsole.properties b/browser/locales/en-US/chrome/browser/devtools/webconsole.properties
--- a/browser/locales/en-US/chrome/browser/devtools/webconsole.properties
+++ b/browser/locales/en-US/chrome/browser/devtools/webconsole.properties
@@ -145,16 +145,21 @@ webConsolePositionBelow=Below
 webConsolePositionWindow=Window
 
 # LOCALIZATION NOTE (webConsoleWindowTitleAndURL): The Web Console floating
 # panel title, followed by the web page URL.
 # For RTL languages you need to set the LRM in the string to give the URL
 # the correct direction.
 webConsoleWindowTitleAndURL=Web Console - %S
 
+# LOCALIZATION NOTE (webConsoleMixedContentWarning): Message displayed after a
+# URL in the Web Console that has been flagged for Mixed Content (i.e. http
+# content in an https page)
+webConsoleMixedContentWarning=Mixed Content
+
 # LOCALIZATION NOTE (scratchpad.linkText):
 # The text used in the right hand side of the web console command line when
 # Javascript is being entered, to indicate how to jump into scratchpad mode
 scratchpad.linkText=Shift+RETURN - Open in Scratchpad
 
 # LOCALIZATION NOTE (gcliterm.instanceLabel): The console displays
 # objects using their type (from the constructor function) in this descriptive
 # string
diff --git a/browser/themes/gnomestripe/devtools/webconsole.css b/browser/themes/gnomestripe/devtools/webconsole.css
--- a/browser/themes/gnomestripe/devtools/webconsole.css
+++ b/browser/themes/gnomestripe/devtools/webconsole.css
@@ -194,16 +194,20 @@
 .webconsole-msg-console.webconsole-msg-warn {
   -moz-image-region: rect(24px, 24px, 32px, 16px);
 }
 
 .webconsole-msg-console.webconsole-msg-info {
   -moz-image-region: rect(24px, 32px, 32px, 24px);
 }
 
+.webconsole-mixed-content {
+  color: #FF0000;
+}
+
 /* Input and output styles */
 .webconsole-msg-input > .webconsole-msg-icon-container,
 .webconsole-msg-output > .webconsole-msg-icon-container {
   border-left: solid #808080 6px;
 }
 
 .webconsole-msg-input {
   -moz-image-region: rect(24px, 40px, 32px, 32px);
diff --git a/browser/themes/pinstripe/devtools/webconsole.css b/browser/themes/pinstripe/devtools/webconsole.css
--- a/browser/themes/pinstripe/devtools/webconsole.css
+++ b/browser/themes/pinstripe/devtools/webconsole.css
@@ -249,16 +249,20 @@
 .webconsole-msg-console.webconsole-msg-warn {
   -moz-image-region: rect(24px, 24px, 32px, 16px);
 }
 
 .webconsole-msg-console.webconsole-msg-info {
   -moz-image-region: rect(24px, 32px, 32px, 24px);
 }
 
+.webconsole-mixed-content {
+  color: #FF0000;
+}
+
 /* Input and output styles */
 .webconsole-msg-input > .webconsole-msg-icon-container,
 .webconsole-msg-output > .webconsole-msg-icon-container {
   border-left: solid #808080 6px;
 }
 
 .webconsole-msg-input {
   -moz-image-region: rect(24px, 40px, 32px, 32px);
diff --git a/browser/themes/winstripe/devtools/webconsole.css b/browser/themes/winstripe/devtools/webconsole.css
--- a/browser/themes/winstripe/devtools/webconsole.css
+++ b/browser/themes/winstripe/devtools/webconsole.css
@@ -200,16 +200,20 @@
 .webconsole-msg-console.webconsole-msg-warn {
   -moz-image-region: rect(24px, 24px, 32px, 16px);
 }
 
 .webconsole-msg-console.webconsole-msg-info {
   -moz-image-region: rect(24px, 32px, 32px, 24px);
 }
 
+.webconsole-mixed-content {
+  color: #FF0000;
+}
+
 /* Input and output styles */
 .webconsole-msg-input > .webconsole-msg-icon-container,
 .webconsole-msg-output > .webconsole-msg-icon-container {
   border-left: solid #808080 6px;
 }
 
 .webconsole-msg-input {
   -moz-image-region: rect(24px, 40px, 32px, 32px);
